Bottom: 3364decc8ef789ea39cb60a8376e3028b6578cf7
Top:    06cda31769fc0713237659ebbbfced12c9d692be
Author: Ian Kent <raven@themaw.net>
Date:   2013-05-03 15:02:07 +0800

Refresh of autofs-5.0.7-use-MS_UNBINDABLE-for-autofs-mounts.patch

---

diff --git a/daemon/direct.c b/daemon/direct.c
index 9d8bcc3..b52f52c 100644
--- a/daemon/direct.c
+++ b/daemon/direct.c
@@ -422,12 +422,18 @@ int do_mount_autofs_direct(struct autofs_point *ap,
 
 	map_name = me->mc->map->argv[0];
 
-	ret = mount(map_name, me->key, "autofs", mountflags, mp->options);
+	ret = mount(map_name, me->key, "autofs", MS_MGC_VAL, mp->options);
 	if (ret) {
 		crit(ap->logopt, "failed to mount autofs path %s", me->key);
 		goto out_err;
 	}
 
+	ret = mount(NULL, me->key, NULL, MS_UNBINDABLE|MS_REC, NULL);
+	if (ret) {
+		warn(ap->logopt,
+		     "failed to set autofs mount unbindable path %s", me->key);
+	}
+
 	ret = stat(me->key, &st);
 	if (ret == -1) {
 		error(ap->logopt,
@@ -770,7 +776,7 @@ int mount_autofs_offset(struct autofs_point *ap, struct mapent *me, const char *
 	if (!type || strcmp(ap->entry->maps->type, "hosts"))
 		map_name = me->mc->map->argv[0];
 
-	ret = mount(map_name, mountpoint, "autofs", mountflags, mp->options);
+	ret = mount(map_name, mountpoint, "autofs", MS_MGC_VAL, mp->options);
 	if (ret) {
 		crit(ap->logopt,
 		     "failed to mount offset trigger %s at %s",
@@ -778,6 +784,13 @@ int mount_autofs_offset(struct autofs_point *ap, struct mapent *me, const char *
 		goto out_err;
 	}
 
+	ret = mount(NULL, mountpoint, NULL, MS_UNBINDABLE|MS_REC, NULL);
+	if (ret) {
+		warn(ap->logopt,
+		     "failed to set mount trigger %s unbindable at %s",
+		     me->key, mountpoint);
+	}
+
 	ret = stat(mountpoint, &st);
 	if (ret == -1) {
 		error(ap->logopt,
